pkg/dio/multi_test.go

```diff
@@ -12,365 +12,278 @@
-		return
-	}
-
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-		return
+	if buffer != expected {
+		t.Errorf("Expected: %q, actual: %q", expected, buffer)
 	}
 }
 
-func TestWriteAdd(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+// TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
+func TestMultiWrite(t *testing.T) {
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
 
-	mwriter := MultiWriter(pipeAW)
+	// create & initialize writers and buffers
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+		reader, writer := io.Pipe()
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+		writers = append(writers, writer)
+		buffers = append(buffers, &buffer)
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data - shouldn't be present
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != "" {
-		t.Errorf("B: expected: %s, actual: %s", "", bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
+}
 
-	// Add writer to existing MultiWriter
-	mwriter.Add(pipeBW)
+// TestMultiWrite creates bunch of multi writers and writes to them, then adds more and writes again
+func TestWriteAdd(t *testing.T) {
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
+
+	var writersAdded []io.Writer
+	var buffersAdded []*bytes.Buffer
+
+	// create & initialize writers and buffers into two categories
+	// *Added ones will be added to multi writer later
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
+
+		reader, writer := io.Pipe()
+
+		// add the writer & buffer to skip list if i divisible by three
+		if i%3 == 0 {
+			writersAdded = append(writersAdded, writer)
+			buffersAdded = append(buffersAdded, &buffer)
+		} else {
+			writers = append(writers, writer)
+			buffers = append(buffers, &buffer)
+		}
 
-	data2 := "verify dynamic add"
-	_, err = mwriter.Write([]byte(data2))
-	if err != nil {
-		t.Error(err)
-		return
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data - shouldn't be present
-	if bufA.String() != data+data2 {
-		t.Errorf("A: expected: %s, actual: %s", data+data2, bufA.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != data2 {
-		t.Errorf("B: expected: %s, actual: %s", data2, bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
```

Comment:
same comment as above about ensuring completion, and for all subsequent control calls.